api/client/credentials.go

```diff
@@ -47,6 +48,13 @@ type Credentials interface {
 	// SSHClientConfig returns SSH configuration used to connect to the
 	// Auth server through a reverse tunnel.
 	SSHClientConfig() (*ssh.ClientConfig, error)
+	// Expiry returns the Credentials expiry if it's possible to know its expiry.
+	// When expiry can be determined returns true, else returns false.
+	// If the Credentials don't expire, returns the zero time.
+	// If the Credential is dynamically refreshed or reloaded, (e.g filesystem
+	// reload or tbot renewal), Expiry returns the expiry of the currently active
+	// Credentials.
+	Expiry() (time.Time, bool)
```

Comment:
I hadn't considered the backwards compatibility of this change before your slack question. I think you're right that we don't want to break custom implementations. We could follow the example of `CredentialsWithDefaultAddrs` below and only check for expiry when the credentials support it. This is also better for credentials that don't have a way to get the expiry like `tlsConfigCreds`.

```
type CredentialsWithExpiry struct {
	Credentials
	// Expiry returns the Credentials expiry if it's possible to know its expiry.
	// When expiry can be determined returns true, else returns false.
	// If the Credentials don't expire, returns the zero time.
	// If the Credential is dynamically refreshed or reloaded, (e.g filesystem
	// reload or tbot renewal), Expiry returns the expiry of the currently active
	// Credentials.
	Expiry() (time.Time, bool)
}
```

Reply:
This is what I did in my initial PR but @espadolini was against this design as it blocks wrapping.

Reply:
What I can do to make everyone happy is backport with the optional interface, and do the breaking change in v17, but that's extra work for me.

Reply:
What do you mean by wrapping? I meant to make the suggesting an interface, not a struct, if that's the issue.

Reply:
Unconditionally implementing the additional method _if it has a way to respond "not implemented/unknown/not applicable"_ is fine.

> What do you mean by wrapping?

If you're only relying on a method being implemented or not to dictate behavior, you end up having to do a bunch of type-level shenanigans if you ever want to write a wrapper over an interface; imagine if you wanted to write a `LoggingCredentials` that writes to the log every time something fetches a TLS or SSH config. If the surrounding code only checked that a `Credentials` interface value implemented `Expiry()` to decide whether or not the expiration time was available, you'd need to do something like ```go
func NewLoggingCredentials(inner Credentials) Credentials {
  if inner, ok := inner.(CredentialsWithExpiry); ok {
    return &loggingCredentialsWithExpiry{inner: inner}
  }
  return &loggingCredentialsWithoutExpiry{inner: inner}
}
```
with the only difference between the two being the type of the inner field (`Credentials` or `CredentialsWithExpiry`).

Reply:
I will merge the PR as-is and will add a new interface in the backport to preserve compatibility within the same major version